File: DailyLeetcode/4Aug/FruitIntoBaskets.cpp
// Optimized Solution
// Time Complexity=O(n) for sliding window
// Space Complexity=O(1) as in map also we store max 2 elements
class Solution
{
public:
    int totalFruit(vector<int> &fruits)
    {
        int n = fruits.size();
        int l = 0; // left pointer
        int r = 0; // right pointer
        int maxl = 0;
        unordered_map<int, int> mp; // make a map  to track two distinct elements
        while (r < n)
        {
            mp[fruits[r]]++;
            while (mp.size() > 2)
            {
                mp[fruits[l]]--;
                if (mp[fruits[l]] == 0)
                { // if freq=0 delete from map
                    mp.erase(fruits[l]);
                }
                l++;
            }
            maxl = max(maxl, r - l + 1);
            r++;
        }
        return maxl;
    }
};

File: DailyLeetcode/4Aug/FruitIntoBaskets.py
from collections import defaultdict

class Solution:
    def totalFruit(self, fruits):
        n = len(fruits)
        l = 0
        r = 0
        max_len = 0
        fruit_count = defaultdict(int)  # works like unordered_map

        while r < n:
            fruit_count[fruits[r]] += 1

            while len(fruit_count) > 2:
                fruit_count[fruits[l]] -= 1
                if fruit_count[fruits[l]] == 0:
                    del fruit_count[fruits[l]]
                l += 1

            max_len = max(max_len, r - l + 1)
            r += 1

        return max_len
